lib/jsdom/browser/Window.js

```diff
@@ -97,6 +154,62 @@ function setupWindow(windowInstance, { runScripts }) {
   mixin(windowInstance, GlobalEventHandlersImpl.prototype);
   windowInstance._initGlobalEvents();
 
+  Object.defineProperty(windowInstance, "onbeforeunload", {
+    configurable: true,
+    enumerable: true,
+    get() {
+      return idlUtils.tryWrapperForImpl(getCurrentEventHandlerValue(this, "beforeunload"));
+    },
+    set(V) {
+      if (!idlUtils.isObject(V)) {
```

Comment:
Why is this necessary? Shouldn't OnBeforeUnloadEventHandlerNonNull handle the [LegacyTreatNonObjectAsNull] ness?

Reply:
The non‑`object` parameter is handled as part of <https://heycam.github.io/webidl/#es-nullable-type>.

Reply:
Sorry, I don't understand. Can you explain more why `OnBeforeUnloadEventHandlerNonNull.convert()` fails to handle non-objects?

Reply:
This is because `OnBeforeUnloadEventHandlerNonNull.convert()` [always returns a `invokeTheCallbackFunction` operation](https://heycam.github.io/webidl/#es-callback-function), because it’s a non‑nullable type.

In order for the conversion to return `null`, it has to be wrapped in a [nullable type conversion](https://heycam.github.io/webidl/#es-nullable-type), where the first step is:

> 1. If [Type](https://tc39.github.io/ecma262/#sec-ecmascript-data-types-and-values)(<var>V</var>) is not Object, and the conversion to an IDL value is being performed due to <var>V</var> being assigned to an [attribute](https://heycam.github.io/webidl/#dfn-attribute) whose type is a [nullable](https://heycam.github.io/webidl/#dfn-nullable-type) [callback function](https://heycam.github.io/webidl/#dfn-callback-function) that is annotated with [[`LegacyTreatNonObjectAsNull`](https://heycam.github.io/webidl/#LegacyTreatNonObjectAsNull)], then return the IDL [nullable type](https://heycam.github.io/webidl/#dfn-nullable-type) <var>T</var>? value **null**.

Reply:
Then is there a spec bug? https://html.spec.whatwg.org/#event-handler-idl-attributes setter algorithm step 3 says "If the given value is null". But you're saying that it will never be null after passing through the IDL layer?

Reply:
Oh, I see, I was confused.

OK, so my question is, why can't we use `OnBeforeUnloadEventHandler.convert(V, ...)`?

Reply:
`OnBeforeUnloadEventHandler`, as a typedef, does not generate conversion files. We could make them generate files, but a majority of them don't get used externally. An webidl2js-specific extended attribute on the typedefs to force-generate the files may be appropriate.

Reply:
Hmm, I see. So if Window was converted to Web IDL, then it would be able to use the typedef, but we can't use it manually?

Reply:
Yes, at least right now.

However, if I were to have to pick the easier of converting `Window` entirely to Web IDL and making typedefs usable outside of Web IDL, I would go for the latter every time :')

Reply:
OK, great, I'm satisfied then. Although I'd like a comment to explain this; I can push a commit that adds that.

I feel like we're relatively close to converting Window to use Web IDL...